### PR TITLE
Initial setup

### PR DESCRIPTION
**Pull request by pierreosselin**
Adding the Gradient Descent Script

### CODE REVIEW

File: gradient_descent.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+function objective(arr) {
+    ret= 0;
+    for (var i = 0, len = arr.length; i < len; i++) {
+        ret = ret + arr[i] * arr[i];
+    }
+    return ret;
+}
+
+class gradient_descent {
```
Comment (Scrabouillmax) [inline gradient_descent.js]:
```suggestion
class GradientDescent {
```

File: gradient_descent.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+const x = [1,2];
+const eps = 0.1;
+
+function objective(arr) {
+    ret= 0;
+    for (var i = 0, len = arr.length; i < len; i++) {
+        ret = ret + arr[i] * arr[i];
+    }
+    return ret;
+}
```
Comment (Scrabouillmax) [inline gradient_descent.js]: You can use `reduce` here:
```suggestion
const objective = x => x.reduce((acc, value) => (acc + value**2), 0);
```

File: gradient_descent.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    for (var i = 0, len = arr.length; i < len; i++) {
+        x_h[i] = arr[i] + this.h;
+        gradient.push((this.objective(x_h) - this.objective(arr))/this.h);
+        x_h[i] = arr[i];
+    }
+    return gradient;
+  }
+
+  one_step(){
+    var gradient = this.differentiate(this.x);
```
Comment (Scrabouillmax) [inline gradient_descent.js]: In general, try to use const or let, not var, it is more precise :)

File: gradient_descent.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  differentiate(arr){
+    var gradient = [];
+    const x_h = arr.map(el => el);
+    for (var i = 0, len = arr.length; i < len; i++) {
+        x_h[i] = arr[i] + this.h;
+        gradient.push((this.objective(x_h) - this.objective(arr))/this.h);
+        x_h[i] = arr[i];
+    }
+    return gradient;
+  }
```
Comment (Scrabouillmax) [inline gradient_descent.js]: You can use map here:
```suggestion
differentiate(arr) {
    const x_h = arr.map(el => el);
    return arr.map((el, i) => {
        x_h[i] = el + this.h;
        gradient = (this.objective(x_h) - this.objective(arr))/this.h);
        x_h[i] = el;
        return gradient;
    })
}
```

File: gradient_descent.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    return norm
+  }
+
+  optimize(eps){
+    var norm = 0;
+    do {
+      norm = (this.one_step())**(1/2)
+      console.log("Norm:", norm, " State:", this.x);
+    } while (norm > eps);
```
Comment (Scrabouillmax) [inline gradient_descent.js:42]: Can we also define a maximum iteration number in case it does not converge?

File: gradient_descent.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    var gradient = this.differentiate(this.x);
+    var norm = 0;
+    for (var i = 0, len = gradient.length; i < len; i++) {
+        this.x[i] = this.x[i] - this.delta * gradient[i];
+        norm = norm + gradient[i]**2
+    }
+    return norm
+  }
+
+  optimize(eps){
```
Comment (Scrabouillmax) [inline gradient_descent.js:37]: I would return the minimum we found here :)

File: gradient_descent.js (hunk 5 above)
Comment (pierreosselin) [reply gradient_descent.js:42]: No

**Pull request merged**